### ISSUE TITLE
Help needed with uploading waveforms into HDAWG

### ISSUE THREAD

**Issue by paniash** [qutech/qupulse#808]
Hi,

Going through the tutorials, I can now make custom pulses as desired. However, I was wondering if there is documentation (perhaps I missed it) on implementing these pulses on the HDAWG? I'd really appreciate any help.

Thanks!

**Comment by terrorfisch** [comment]
Hi,

currently, the master branch driver only supports the old version of the HDAWG firmware since ZI removed the `playWaveIndexed` function call. Sadly this version is only available for python 3.8. We have a driver for the new version in development which we plan to release this month.

qupulse provides drivers for uploading instantiated pulse templates to AWGs under `qupulse.hardware.awgs`. However, these are a bit cumbersome to use as they are not designed for direct user interaction. You can use it as a building block to drive your experiment. There is the `HarwareSetup` class in `qupulse.hardware.setup` which integrates multiple awgs and data acquisition devices which we use.

The following code snippet requires `zhinst<=20.7.2701` and a corresponding firmware version

```
from qupulse.hardware.setup import HardwareSetup, PlaybackChannel, MarkerChannel
from qupulse.hardware.awgs.import HDAWGRepresentation, HDAWGChannelGrouping
hardware_setup = HardwareSetup()

# physical device
your_hdawg = HDAWGRepresentation('DEV0000', ...)

# grouped channels
your_hdawg.channel_grouping = HDAWGChannelGrouping.CHAN_GROUP_1x8
all_channels, = your_hdawg.channel_tuples

# Map left plunger of hardware setup to first channel of channel group
hardware_setup.set_channel('LeftPlunger', PlaybackChannel(all_channels, 0))
hardware_setup.set_channel('RightPlunger', PlaybackChannel(all_channels, 1))
```

Now you can use the harware setup and upload programs i.e. instantiated pulses

```python
# has channels L and R
your_pulse_template = ...

channel_mapping = {'L': 'LeftPlunger', 'R': 'RightPlunger'}
your_program = your_pulse_template.create_program(..., channel_mapping=channel_mapping)

hardware_setup.register_program('meaningful_program_name', your_program,
    # this is a callback handle that triggers the program. Needs adjustment if you trigger the AWG externally
    all_channels.run_current_program)
```

And upload and run
```python
hardware_setup.run_program('meaningful_program_name')
```

I have in on my to do list to create a proper example for this...

**Comment by terrorfisch** [comment]
@paniash You can try https://github.com/qutech/qupulse-hdawg Currently, the documentation is missing and there might be unexpected problems due to a very fresh refector. Please let us know if you encounter any.

**Comment by paniash** [comment]
Thanks!

**Comment by terrorfisch** [comment]
Implemented by #819
